server/join.go

```diff
@@ -37,40 +43,113 @@ func (cfg *Config) genClientV3Config() clientv3.Config {
 	}
 }
 
-// prepareJoinCluster send MemberAdd command to pd cluster,
-// returns pd initial cluster configuration.
-func (cfg *Config) prepareJoinCluster() (string, error) {
+func memberAdd(client *clientv3.Client, urls []string) (*clientv3.MemberAddResponse, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), defaultDialTimeout)
 	defer cancel()
 
+	return client.MemberAdd(ctx, urls)
+}
+
+func memberList(client *clientv3.Client) (*clientv3.MemberListResponse, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultDialTimeout)
+	defer cancel()
+
+	return client.MemberList(ctx)
+}
+
+func (cfg *Config) hasPreviousData() bool {
```

Comment:
config now has too many functions for cluster handling, you should move all to another structure.